ambry-router/src/test/java/com.github.ambry.router/PutOperationTest.java

```diff
@@ -0,0 +1,180 @@
+import com.github.ambry.network.RequestInfo;
+import com.github.ambry.network.ResponseInfo;
+import com.github.ambry.protocol.PutRequest;
+import com.github.ambry.protocol.RequestOrResponse;
+import com.github.ambry.utils.ByteBufferChannel;
+import com.github.ambry.utils.MockTime;
+import com.github.ambry.utils.Time;
+import com.github.ambry.utils.Utils;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Random;
+import java.util.TreeMap;
+import org.junit.Assert;
+import org.junit.Test;
+
+
+public class PutOperationTest {
+  private final RouterConfig routerConfig;
+  private final MockClusterMap mockClusterMap;
+  private final ResponseHandler responseHandler;
+  private final Time time;
+  private final Map<Integer, PutOperation> correlationIdToPutOperation = new TreeMap<>();
+  private final MockServer mockServer = new MockServer();
+
+  private class PutTestRequestRegistrationCallbackImpl implements PutRequestRegistrationCallback {
+    private List<RequestInfo> requestListToFill;
+
+    @Override
+    public void registerRequestToSend(PutOperation putOperation, RequestInfo requestInfo) {
+      requestListToFill.add(requestInfo);
+      correlationIdToPutOperation.put(((RequestOrResponse) requestInfo.getRequest()).getCorrelationId(), putOperation);
+    }
+  }
+
+  ;
+  private final PutTestRequestRegistrationCallbackImpl requestRegistrationCallback =
+      new PutTestRequestRegistrationCallbackImpl();
+
+  private final int chunkSize = 10;
+  private final int requestParallelism = 3;
+  private final int successTarget = 1;
+  private final Random random = new Random();
+
+  public PutOperationTest()
+      throws Exception {
+    Properties properties = new Properties();
+    properties.setProperty("router.hostname", "localhost");
+    properties.setProperty("router.datacenter.name", "DC1");
+    properties.setProperty("router.max.put.chunk.size.bytes", Integer.toString(chunkSize));
+    properties.setProperty("router.put.request.parallelism", Integer.toString(requestParallelism));
+    properties.setProperty("router.put.success.target", Integer.toString(successTarget));
+    VerifiableProperties vProps = new VerifiableProperties(properties);
+    routerConfig = new RouterConfig(vProps);
+    mockClusterMap = new MockClusterMap();
+    responseHandler = new ResponseHandler(mockClusterMap);
+    time = new MockTime();
+  }
+
+  /**
+   * Ensure that if any of the requests associated with the buffer of a PutChunk is not completely read out even
+   * after the associated chunk is complete, the buffer is not reused even though the PutChunk is reused.
+   */
+  @Test
+  public void testSendIncomplete()
+      throws Exception {
+    BlobProperties blobProperties =
+        new BlobProperties(chunkSize * 5, "serviceId", "memberId", "contentType", false, Utils.Infinite_Time);
```

Comment:
would prefer if we used the newly created constants for max in memory chunks (is that in the other PR?) here. That way, if we were to tune/change that, these tests wouldn't suddenly fail.